R/Previewer.R

```diff
@@ -124,122 +124,124 @@ reporter_previewer_srv <- function(id,
     )
     report_load_srv("load", reporter = reporter)
     reset_report_button_srv("reset", reporter = reporter)
+    reporter_previewer_cards_srv("cards", reporter)
+  })
+}
+
+reporter_previewer_cards_ui <- function(id) {
+  ns <- shiny::NS(id)
+  shiny::tags$div(
+    id = "reporter_previewer",
+    tags$div(
+      id = ns("empty_reporters"),
+      tags$h4(
+        class = "text-muted",
+        shiny::icon("circle-info"),
+        "No reports have been added yet."
+      )
+    ),
+    bslib::accordion(id = ns("reporter_cards"), open = FALSE)
+  )
+}
+
+reporter_previewer_cards_srv <- function(id, reporter) {
+  moduleServer(id, function(input, output, session) {
+    current_cards_rvs <- shiny::reactiveValues()
 
-    current_cards <- shiny::reactiveVal()
-    insert_cards <- shiny::reactiveVal()
-    remove_cards <- shiny::reactiveVal()
+    insert_queue_rv <- shiny::reactiveVal()
+    remove_queue_rv <- shiny::reactiveVal()
     shiny::observeEvent(reporter$get_reactive_add_card(), {
-      reporter_ids <- vapply(reporter$get_cards(), attr, character(1L), which = "id")
-      current_ids <- vapply(current_cards(), attr, character(1L), which = "id")
-
-      to_add <- reporter$get_cards()[!reporter_ids %in% current_ids]
-      to_remove <- current_cards()[!current_ids %in% reporter_ids]
-      if (length(to_add)) insert_cards(to_add)
-      if (length(to_remove)) remove_cards(to_remove)
-      current_cards(reporter$get_cards())
+      all_cards <- reporter$get_cards()
+      reporter_ids <- names(all_cards)
+      current_cards <- Filter(Negate(is.null), shiny::reactiveValuesToList(current_cards_rvs))
+      current_ids <- names(current_cards)
+
+      # Update modified card (there is no support for ReportCard implementation)
+      lapply(
+        reporter_ids[
+          !vapply(all_cards, rlang::obj_address, character(1)) %in%
+            vapply(current_cards, rlang::obj_address, character(1))
+        ],
+        function(modified_id) {
+          current_cards_rvs[[modified_id]] <- all_cards[[modified_id]]
+          NULL
+        }
+      )
+
+      to_add <- !reporter_ids %in% current_ids
+      to_remove <- !current_ids %in% reporter_ids
+      if (any(to_add)) insert_queue_rv(reporter_ids[to_add])
+      if (any(to_remove)) remove_queue_rv(current_ids[to_remove])
     })
 
-    shiny::observeEvent(insert_cards(), {
-      cards <- insert_cards()
-      lapply(names(cards), function(card_name) {
-        card_id <- attr(cards[[card_name]], "id", exact = TRUE)
+    shiny::observeEvent(insert_queue_rv(), {
+      lapply(insert_queue_rv(), function(card_id) {
         bslib::accordion_panel_insert(
           id = "reporter_cards",
-          reporter_previewer_card_ui(id = session$ns(card_id), card_name = card_name)
+          reporter_previewer_card_ui(id = session$ns(card_id), card_id = card_id)
         )
+        current_cards_rvs[[card_id]] <- reporter$get_cards(card_id)[[1]]
```

Comment:
Change in `$get_cards` to introduce an index is not needed. Returning a list is good enough to not introduce new lines of code in `$get_cards`. `get_cards(index)` is used here only  


```suggestion
        current_cards_rvs[[card_id]] <- reporter$get_cards()[[card_id]]
```